File: leet-code/kClosest.js
/**
 * @param {number[][]} points
 * @param {number} K
 * @return {number[][]}
 */
var kClosest = function (points, K) {
  points = points.sort((a, b) => {
    const tempA = Math.pow(a[0], 2) + Math.pow(a[1], 2)
    const tempB = Math.pow(b[0], 2) + Math.pow(b[1], 2)
    return tempA - tempB
  })
  return points.slice(0, K)
}

// console.log(
//   kClosest(
//     [
//       [3, 3],
//       [5, -1],
//       [-2, 4]
//     ],
//     2
//   )
// )

// 用快排看看会不会更快

var kClosest = function (points, K) {
  points = quickSort(points, 0, points.length - 1)
  return points.slice(0, K)
}

function quickSort (points, left, right) {
  if (left >= right) {
    return
  }
  let i = left
  let j = right
  let flag = points[left]
  while (i < j) {
    while (i < j && definePow(points[j]) >= definePow(flag)) {
      j--
    }
    while (i < j && definePow(points[i]) <= definePow(flag)) {
      i++
    }
    let temp = points[i]
    points[i] = points[j]
    points[j] = temp
  }
  points[left] = points[j]
  points[j] = flag
  quickSort(points, left, j - 1)
  quickSort(points, j + 1, right)
  return points
}

function definePow (arr) {
  return Math.pow(arr[0], 2) + Math.pow(arr[1], 2)
}
console.log(
  kClosest(
    [
      [3, 3],
      [5, -1],
      [-2, 4]
    ],
    2
  )
)

File: leet-code/kClosest.go
package main

import "sort"

// Points is
type Points [][]int

// Len is
func (p Points) Len() int { return len(p) }

// Swap is
func (p Points) Swap(i, j int) { p[i], p[j] = p[j], p[i] }

// Less is
func (p Points) Less(i, j int) bool {
	return squareAbs(p[i][0])+squareAbs(p[i][1]) < squareAbs(p[j][0])+squareAbs(p[j][1])
}

func kClosest(points [][]int, K int) [][]int {
	sort.Sort(Points(points))
	return points[:K]
}

func squareAbs(i int) int {
	if i < 0 {
		i = -i
	}
	return i * i
}

// golang的sort自定义之后效率好像不得行。。手动排序试试

func kClosest2(points [][]int, K int) [][]int {
	kClosestQuickSort(points, 0, len(points)-1)
	return points[:K]
}

func kClosestQuickSort(points [][]int, left, right int) [][]int {
	if left >= right {
		return points
	}

	i, j := left, right
	flag := points[left]

	for i < j {
		for i < j && squareAbsPoints(points[j]) >= squareAbsPoints(flag) {
			j--
		}
		for i < j && squareAbsPoints(points[i]) <= squareAbsPoints(flag) {
			i++
		}
		temp := points[i]
		points[i] = points[j]
		points[j] = temp
	}
	points[left] = points[j]
	points[j] = flag
	kClosestQuickSort(points, left, j-1)
	kClosestQuickSort(points, j+1, right)
	return points
}

func squareAbsPoints(point []int) int {
	i := point[0]
	j := point[1]

	if i < 0 {
		i = -i
	}
	if j < 0 {
		j = -j
	}

	return i*i + j*j
}
